include/deal.II/multigrid/mg_transfer_global_coarsening.h

```diff
@@ -0,0 +1,444 @@
+                   VectorType &      dst,
+                   const VectorType &src) const;
+};
+
+
+
+/**
+ * Class for transfer between two multigrid levels for p or global coarsening.
+ * Specialization for LinearAlgebra::distributed::Vector.
+ */
+template <int dim, typename Number>
+class MGTwoLevelTransfer<dim, LinearAlgebra::distributed::Vector<Number>>
+{
+public:
+  /**
+   * Set up global coarsening transfer @p transfer between the given
+   * dof-handlers.
+   */
+  void
+  reinit_geometric_transfer(const DoFHandler<dim> &          dof_handler_fine,
+                            const DoFHandler<dim> &          dof_handler_coarse,
+                            const AffineConstraints<Number> &constraint_fine,
+                            const AffineConstraints<Number> &constraint_coarse);
+
+  /**
+   * Set up polynomial coarsening transfer @p transfer between the given
+   * dof-handlers.
+   *
+   * @note The function polynomial_transfer_supported() can be used to
+   *   check if the given polynomial coarsening strategy is supported.
+   */
+  void
+  reinit_polynomial_transfer(
+    const DoFHandler<dim> &          dof_handler_fine,
+    const DoFHandler<dim> &          dof_handler_coarse,
+    const AffineConstraints<Number> &constraint_fine,
+    const AffineConstraints<Number> &constraint_coarse);
+
+  /**
+   * Check if the polynomial transfer between @p fe_degree_fine and
+   * @p fe_degree_coarse is supported.
+   *
+   * @note Currently, the polynomial coarsening strategies: 1) go-to-one,
+   *   2) bisect, and 3) decrease-by-one is supported.
+   */
+  static bool
+  polynomial_transfer_supported(const unsigned int fe_degree_fine,
+                                const unsigned int fe_degree_coarse);
+
+  /**
+   * Perform prolongation.
+   */
+  void
+  prolongate(const unsigned int,
+             LinearAlgebra::distributed::Vector<Number> &      dst,
+             const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+  /**
+   * Perform restriction.
+   */
+  void
+  restrict_and_add(const unsigned int,
+                   LinearAlgebra::distributed::Vector<Number> &      dst,
+                   const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+  /**
+   * Print internal data structures to stream @p out.
+   */
+  template <typename Stream>
+  void
+  print_internal(Stream &out) const;
+
+private:
+  /**
+   * A multigrid transfer scheme. A multrigrid transfer class can have different
+   * transfer schemes to enable p-adaptivity (one transfer scheme per
+   * polynomial degree pair) and to enable global coarsening (one transfer
+   * scheme for transfer between children and parent cells, as well as, one
+   * transfer scheme for cells that are not refined).
+   */
+  struct MGTransferScheme
+  {
+    /**
+     * Number of coarse cells.
+     */
+    unsigned int n_coarse_cells;
+
+    /**
+     * Number of degrees of freedom of a coarse cell.
+     */
+    unsigned int n_dofs_coarse_cell;
```

Comment:
I am not sure about the name. Do you mean
```suggestion
    unsigned int dofs_per_cell_coarse;
```
(and similarly for `n_dofs_fine_cell`). The reason why I'm asking is that I think of a global quantity with `n_dofs`, whereas this seems local.